This project is to visualise and analyse the gender gap in the tech industry, especially in the Data Science field. There are only arond 19,000 respondents, and that this sample population is not representative of the general population. However, this will give us a good idea of the actual population.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.graph_objs as go
import plotly.offline as py
import math

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/loycatherine/dataset/master/multiple_choice_responses.csv')
df = df.drop(df.index[0])

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning:

Columns (0,3,7,19,34,47,49,50,51,52,53,54,68,81,94,96,109,115,130,139,147,154,167,180,193,206,219,232,245) have mixed types.Specify dtype option on import or set low_memory=False.



##Data cleaning

In [3]:
df = df[['Q1','Q2','Q3','Q4','Q5','Q10']]
df.columns = ["age", "gender", "country", "education", "job title", "salary"]

In [4]:
df.dtypes

age          object
gender       object
country      object
education    object
job title    object
salary       object
dtype: object

In [5]:
df.loc[df.salary == '$0-999', 'salary'] = '0-999'
df.loc[df.salary == '> $500,000', 'salary'] = '> 500,000'

In [6]:
df

age  gender  ...          job title           salary
1      22-24    Male  ...  Software Engineer    30,000-39,999
2      40-44    Male  ...  Software Engineer      5,000-7,499
3      55-59  Female  ...                NaN              NaN
4      40-44    Male  ...              Other  250,000-299,999
5      22-24    Male  ...              Other      4,000-4,999
...      ...     ...  ...                ...              ...
19713  50-54    Male  ...                NaN              NaN
19714  18-21    Male  ...              Other            0-999
19715  35-39    Male  ...            Student              NaN
19716  25-29    Male  ...       Statistician      1,000-1,999
19717  50-54    Male  ...  Software Engineer    60,000-69,999

[19717 rows x 6 columns]

In [7]:
job_title = df.groupby("job title")

##Visualise Gender Distribution

In [8]:
df.gender.unique()

array(['Male', 'Female', 'Prefer to self-describe', 'Prefer not to say'],
      dtype=object)

In [9]:
gender = df['gender'].value_counts()
labels = gender.index
percentages = np.array((gender / gender.sum())*100)
color_palette_list = ['#A5C8E4', '#C0ECCC', '#F9F0C1', '#F6A8A6']
trace = go.Pie(labels=labels, 
               hoverinfo='label+percent', 
               values=percentages, 
               textposition='outside',
               marker=dict(colors=color_palette_list),
               rotation=90)

layout = go.Layout(
                    title="Gender Distribution",
                    font=dict(family='Arial', size=12, color='#909090'),
                    legend=dict(x=0.9, y=0.5)
                    )
data = [trace]
fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='Gender Distribution')

As you can see, 81% of the respondents are male, and 16% of the respondents are female. This shows the predominance of males in the tech industry.

##Visualise Age Distribution

In [10]:
age = df['age'].value_counts()
age.index

Index(['25-29', '22-24', '30-34', '18-21', '35-39', '40-44', '45-49', '50-54',
       '55-59', '60-69', '70+'],
      dtype='object')

In [11]:
np.array(age)

array([4458, 3610, 3120, 2502, 2087, 1439,  949,  692,  422,  338,  100])

In [12]:
age.index

Index(['25-29', '22-24', '30-34', '18-21', '35-39', '40-44', '45-49', '50-54',
       '55-59', '60-69', '70+'],
      dtype='object')

In [13]:
labels = age.index
percentages = np.array((age / age.sum())*100)
trace = go.Pie(labels=labels, 
               hoverinfo='label+percent', 
               values=percentages, 
               textposition='outside',
               rotation=90)

layout = go.Layout(
                    title="Age Distribution",
                    font=dict(family='Arial', size=12, color='#909090'),
                    legend=dict(x=0.9, y=0.5)
                    )
data = [trace]
fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='Age Distribution')

From the pie chart, most of the respondents are quite young, around 20-30 years of age. This could be due to the technilogical boom happening around this period of time, encouraging more people to join the tech industry.

##Visualise Education Level Distribution

In [14]:
education = df['education'].value_counts()

colors_bar = ['#111258', '#4C1559','#B1235E', '#F0855F', '#F8C48A' ,'#F8F4A9', '#faf8dc']

trace = go.Bar(
            x=education.index,
            y=np.array(education),
            marker=dict(
            color=colors_bar))

data = [trace]

layout = go.Layout(
    title='Education Distribution',
    font=dict(color='#909090'),
    xaxis=dict(
        title='Education Level',
        titlefont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
        showticklabels=True,
        tickangle=-45,
        tickfont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
),
    yaxis=dict(
        title="No. of respondants with particular education level",
        titlefont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
        showticklabels=True,
        tickangle=0,
        tickfont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        )
    )
)
fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='education distribution')

Most of the respondents have a Master's Degree, or at least a Bachelor's degree

##Visualise Education level distribution by Gender

In [15]:
education_gender = df.groupby(['gender', 'education']).agg({'education':'count'})
education_gender

education
gender                  education                                                    
Female                  Bachelor’s degree                                         865
                        Doctoral degree                                           521
                        I prefer not to answer                                     50
                        Master’s degree                                          1496
                        No formal education past high school                       26
                        Professional degree                                       103
                        Some college/university study without earning a...         87
Male                    Bachelor’s degree                                        5049
                        Doctoral degree                                          2186
                        I prefer not to answer                                    236
                        Master’s degree                                          6930
                        No formal education past high school                      199
                        Professional degree                                       496
                        Some college/university study without earning a...        728
Prefer not to say       Bachelor’s degree                                          72
                        Doctoral degree                                            48
                        I prefer not to answer                                     44
                        Master’s degree                                           106
                        No formal education past high school                        7
                        Professional degree                                         9
                        Some college/university study without earning a...         18
Prefer to self-describe Bachelor’s degree                                           7
                        Doctoral degree                                            12
                        I prefer not to answer                                      3
                        Master’s degree                                            17
                        No formal education past high school                        1
                        Professional degree                                         3
                        Some college/university study without earning a...          4

In [63]:
labels = education_gender.loc["Female"].index
percentages = np.array((education_gender.loc["Female"]["education"] / gender.Female.sum())*100)
# colors_bar = ['#111258', '#4C1559','#B1235E', '#F0855F', '#F8C48A' ,'#F8F4A9', '#faf8dc']
trace = go.Pie(labels=labels, 
               hoverinfo='label+percent', 
               values=percentages, 
               textposition='outside',
               rotation=90)

layout = go.Layout(
                    title="Education Distribution for Females",
                    font=dict(family='Arial', size=12, color='#909090'),
                    legend=dict(x=0.9, y=0.5)
                    )
data = [trace]
fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='Education Distribution for Females')

In [62]:
labels = education_gender.loc["Male"].index
percentages = np.array((education_gender.loc["Male"]["education"] / gender.Male.sum())*100)
trace = go.Pie(labels=labels, 
               hoverinfo='label+percent', 
               values=percentages, 
               textposition='outside',
               rotation=90)

layout = go.Layout(
                    title="Education Distribution for Males",
                    font=dict(family='Arial', size=12, color='#909090'),
                    legend=dict(x=0.9, y=0.5)
                    )
data = [trace]
fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='Education Distribution for Males')

Generally, both males and females are equally educated.

##Visualise Job title Distribution

In [17]:
job_title = df['job title'].value_counts()
job_title

Data Scientist             4085
Student                    4014
Software Engineer          2705
Other                      1690
Data Analyst               1598
Research Scientist         1470
Not employed                942
Business Analyst            778
Product/Project Manager     723
Data Engineer               624
Statistician                322
DBA/Database Engineer       156
Name: job title, dtype: int64

In [18]:
colors_bar = ['#9fcc47', '#4ac9e2' , '#e045c1' , '#efcb5d' , '#f4a69f' , '#63c92c' , '#7da6e8' , '#ffc9b2' ,'#f7b7cb', '#510d91', '#b5520c', '#021982']

trace = go.Bar(
            x=job_title.index,
            y=np.array(job_title),
            marker=dict(
            color=colors_bar))

data = [trace]

layout = go.Layout(
    title='Job Title Distribution',
    font=dict(color='#909090'),
    xaxis=dict(
        title='Job Titles',
        titlefont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
        showticklabels=True,
        tickangle=-45,
        tickfont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
  	),
    yaxis=dict(
        title="No. of respondents",
        titlefont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
        showticklabels=True,
        tickangle=0,
        tickfont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        )
    )
)
fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='job title distribution')

Most of the respondents are either Data Scientists or Students. This is expected as the dataset is from Kaggle, where many students and data scientists visit. 

##Visualise Gender distribution for each Job

In [19]:
job_gender = df.groupby(['gender', 'job title']).agg({'gender':'count'})

In [20]:
job_gender

gender
gender                  job title                      
Female                  Business Analyst            132
                        DBA/Database Engineer        17
                        Data Analyst                347
                        Data Engineer                76
                        Data Scientist              606
                        Not employed                214
                        Other                       258
                        Product/Project Manager      83
                        Research Scientist          241
                        Software Engineer           313
                        Statistician                 63
                        Student                     764
Male                    Business Analyst            634
                        DBA/Database Engineer       136
                        Data Analyst               1231
                        Data Engineer               534
                        Data Scientist             3431
                        Not employed                699
                        Other                      1387
                        Product/Project Manager     627
                        Research Scientist         1198
                        Software Engineer          2338
                        Statistician                256
                        Student                    3186
Prefer not to say       Business Analyst             11
                        DBA/Database Engineer         3
                        Data Analyst                 20
                        Data Engineer                12
                        Data Scientist               42
                        Not employed                 22
                        Other                        34
                        Product/Project Manager      13
                        Research Scientist           28
                        Software Engineer            47
                        Statistician                  2
                        Student                      57
Prefer to self-describe Business Analyst              1
                        Data Engineer                 2
                        Data Scientist                6
                        Not employed                  7
                        Other                        11
                        Research Scientist            3
                        Software Engineer             7
                        Statistician                  1
                        Student                       7

In [21]:
gender = df['gender'].value_counts()
gender

Male                       16138
Female                      3212
Prefer not to say            318
Prefer to self-describe       49
Name: gender, dtype: int64

In [58]:
labels = job_gender.loc["Female"].index
percentages = np.array((job_gender.loc["Female"]["gender"] / gender.Female.sum())*100)
colors_bar = ['#9fcc47', '#4ac9e2' , '#e045c1' , '#efcb5d' , '#f4a69f' , '#63c92c' , '#7da6e8' , '#ffc9b2' ,'#f7b7cb', '#510d91', '#b5520c', '#021982']
trace = go.Pie(labels=labels, 
               hoverinfo='label+percent', 
               values=percentages, 
               textposition='outside',
               rotation=90)

layout = go.Layout(
                    title="Job Distribution for Females",
                    font=dict(family='Arial', size=12, color='#909090'),
                    legend=dict(x=0.9, y=0.5)
                    )
data = [trace]
fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='Job Distribution for Females')

In [59]:
labels = job_gender.loc["Female"].index
percentages = np.array((job_gender.loc["Male"]["gender"] / gender.Male.sum())*100)
colors_bar = ['#9fcc47', '#4ac9e2' , '#e045c1' , '#efcb5d' , '#f4a69f' , '#63c92c' , '#7da6e8' , '#ffc9b2' ,'#f7b7cb', '#510d91', '#b5520c', '#021982']
trace = go.Pie(labels=labels, 
               hoverinfo='label+percent', 
               values=percentages, 
               textposition='outside',
               rotation=90)

layout = go.Layout(
                    title="Job Distribution for Males",
                    font=dict(family='Arial', size=12, color='#909090'),
                    legend=dict(x=0.9, y=0.5)
                    )
data = [trace]
fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='Job Distribution for Males')

##Visualise Salary Distribution

In [24]:
salary = df["salary"].value_counts().reset_index()
salary.columns = ["salary", "count"]
salary["int salary"] = ""

for idx in salary.index:
  salary["int salary"][idx] = int(salary["salary"][idx].replace(',', '').replace('> ', '').split('-')[0])

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [25]:
salary

salary  count int salary
0             0-999   1513          0
1     10,000-14,999    833      10000
2   100,000-124,999    750     100000
3     30,000-39,999    728      30000
4     40,000-49,999    719      40000
5     50,000-59,999    704      50000
6       1,000-1,999    599       1000
7     60,000-69,999    576      60000
8       5,000-7,499    536       5000
9     15,000-19,999    529      15000
10    20,000-24,999    526      20000
11    70,000-79,999    524      70000
12  125,000-149,999    483     125000
13    25,000-29,999    482      25000
14  150,000-199,999    434     150000
15      7,500-9,999    408       7500
16    80,000-89,999    405      80000
17      2,000-2,999    390       2000
18    90,000-99,999    377      90000
19      3,000-3,999    305       3000
20      4,000-4,999    289       4000
21  200,000-249,999    165     200000
22        > 500,000     83     500000
23  300,000-500,000     74     300000
24  250,000-299,999     65     250000

In [26]:
sort_salary = salary.sort_values(by=['int salary'])
sort_salary.drop(columns=['int salary'], inplace=True)

In [27]:
sort_salary

salary  count
0             0-999   1513
6       1,000-1,999    599
17      2,000-2,999    390
19      3,000-3,999    305
20      4,000-4,999    289
8       5,000-7,499    536
15      7,500-9,999    408
1     10,000-14,999    833
9     15,000-19,999    529
10    20,000-24,999    526
13    25,000-29,999    482
3     30,000-39,999    728
4     40,000-49,999    719
5     50,000-59,999    704
7     60,000-69,999    576
11    70,000-79,999    524
16    80,000-89,999    405
18    90,000-99,999    377
2   100,000-124,999    750
12  125,000-149,999    483
14  150,000-199,999    434
21  200,000-249,999    165
24  250,000-299,999     65
23  300,000-500,000     74
22        > 500,000     83

In [28]:
trace = go.Bar(
            x=np.array(sort_salary["salary"]),
            y=np.array(sort_salary["count"]),
            marker=dict(
            color='#746bf4'))

data = [trace]

layout = go.Layout(
    title='Salary Distribution',
    font=dict(color='#909090'),
    xaxis=dict(
        title='Salary',
        titlefont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
        showticklabels=True,
        tickangle=-45,
        tickfont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
),
    yaxis=dict(
        title="No. of respondants with said salary",
        titlefont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
        showticklabels=True,
        tickangle=0,
        tickfont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        )
    )
)
fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='job title distribution')

Interestingly, most of the respondents only earn between $0-999. This could be attributed to the fact that many of the respondents are students, hence they are not earning an income.

In [29]:
counter = 0
salary_map = {}
for idx in sort_salary.index:
  salary_map[sort_salary['salary'][idx]] = counter
  counter += 1

In [30]:
salary_map

{'0-999': 0,
 '1,000-1,999': 1,
 '10,000-14,999': 7,
 '100,000-124,999': 18,
 '125,000-149,999': 19,
 '15,000-19,999': 8,
 '150,000-199,999': 20,
 '2,000-2,999': 2,
 '20,000-24,999': 9,
 '200,000-249,999': 21,
 '25,000-29,999': 10,
 '250,000-299,999': 22,
 '3,000-3,999': 3,
 '30,000-39,999': 11,
 '300,000-500,000': 23,
 '4,000-4,999': 4,
 '40,000-49,999': 12,
 '5,000-7,499': 5,
 '50,000-59,999': 13,
 '60,000-69,999': 14,
 '7,500-9,999': 6,
 '70,000-79,999': 15,
 '80,000-89,999': 16,
 '90,000-99,999': 17,
 '> 500,000': 24}

In [31]:
df["salary map"] = ""
df["salary map"] = df['salary'].map(salary_map)

In [32]:
df

age  gender    country  ...          job title           salary salary map
1      22-24    Male     France  ...  Software Engineer    30,000-39,999       11.0
2      40-44    Male      India  ...  Software Engineer      5,000-7,499        5.0
3      55-59  Female    Germany  ...                NaN              NaN        NaN
4      40-44    Male  Australia  ...              Other  250,000-299,999       22.0
5      22-24    Male      India  ...              Other      4,000-4,999        4.0
...      ...     ...        ...  ...                ...              ...        ...
19713  50-54    Male      Japan  ...                NaN              NaN        NaN
19714  18-21    Male      India  ...              Other            0-999        0.0
19715  35-39    Male      India  ...            Student              NaN        NaN
19716  25-29    Male      India  ...       Statistician      1,000-1,999        1.0
19717  50-54    Male     France  ...  Software Engineer    60,000-69,999       14.0

[19717 rows x 7 columns]

In [33]:
salary_job = df.groupby('job title').agg({'salary map':'median'})

In [34]:
salary_job

salary map
job title                          
Business Analyst                9.0
DBA/Database Engineer          11.0
Data Analyst                    8.0
Data Engineer                  11.0
Data Scientist                 12.0
Not employed                    NaN
Other                          10.0
Product/Project Manager        12.0
Research Scientist             10.0
Software Engineer               8.0
Statistician                    9.0
Student                         NaN

In [35]:
swap_salary_map = dict([(value, key) for key, value in salary_map.items()]) 

In [36]:
swap_salary_map

{0: '0-999',
 1: '1,000-1,999',
 2: '2,000-2,999',
 3: '3,000-3,999',
 4: '4,000-4,999',
 5: '5,000-7,499',
 6: '7,500-9,999',
 7: '10,000-14,999',
 8: '15,000-19,999',
 9: '20,000-24,999',
 10: '25,000-29,999',
 11: '30,000-39,999',
 12: '40,000-49,999',
 13: '50,000-59,999',
 14: '60,000-69,999',
 15: '70,000-79,999',
 16: '80,000-89,999',
 17: '90,000-99,999',
 18: '100,000-124,999',
 19: '125,000-149,999',
 20: '150,000-199,999',
 21: '200,000-249,999',
 22: '250,000-299,999',
 23: '300,000-500,000',
 24: '> 500,000'}

In [37]:
salary_job["median salary range"] = ""
salary_job["median salary range"] = salary_job['salary map'].map(swap_salary_map)

In [38]:
salary_job = salary_job[salary_job['median salary range'].notna()]

In [39]:
salary_job

salary map median salary range
job title                                              
Business Analyst                9.0       20,000-24,999
DBA/Database Engineer          11.0       30,000-39,999
Data Analyst                    8.0       15,000-19,999
Data Engineer                  11.0       30,000-39,999
Data Scientist                 12.0       40,000-49,999
Other                          10.0       25,000-29,999
Product/Project Manager        12.0       40,000-49,999
Research Scientist             10.0       25,000-29,999
Software Engineer               8.0       15,000-19,999
Statistician                    9.0       20,000-24,999

In [40]:
colors_bar = ['#9fcc47', '#4ac9e2' , '#e045c1' , '#efcb5d' , '#f4a69f' , '#63c92c' , '#7da6e8' , '#ffc9b2' ,'#f7b7cb', '#510d91', '#b5520c', '#021982']
trace = go.Bar(
            x=salary_job.index,
            y=np.array(salary_job["salary map"]),
            marker=dict(
            color=colors_bar))

data = [trace]

layout = go.Layout(
    title='Median Salary of Job',
    font=dict(color='#909090'),
    xaxis=dict(
        title='Job Title',
        titlefont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
        showticklabels=True,
        tickangle=-45,
        tickfont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
),
    yaxis=dict(
        title="Median Salary Range",
        titlefont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
        tickmode = 'array',
        tickvals = np.array(list(swap_salary_map.keys())),
        ticktext = np.array(list(swap_salary_map.values())),
        tickangle=0,
        tickfont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        )
    )
)
fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='salary by job')

From the bar graph above, we can see that Data Analyst and Software Engineers have the lowest median salary range, between 15,000 to 19,999. On the other hand, Data Scientists and Product/Project Managers have the highest median salary range, between 40,000 to 49,000.

##Visualise Salary distribution by Gender

In [41]:
salary_job_gender = df.groupby(['gender', 'job title']).agg({'salary map':'median'})

In [42]:
salary_job_gender

salary map
gender                  job title                          
Female                  Business Analyst                7.0
                        DBA/Database Engineer          10.0
                        Data Analyst                    7.0
                        Data Engineer                   7.0
                        Data Scientist                 11.0
                        Not employed                    NaN
                        Other                           7.0
                        Product/Project Manager        11.0
                        Research Scientist              8.0
                        Software Engineer               7.0
                        Statistician                    4.5
                        Student                         NaN
Male                    Business Analyst                9.0
                        DBA/Database Engineer          11.0
                        Data Analyst                    8.0
                        Data Engineer                  11.0
                        Data Scientist                 12.0
                        Not employed                    NaN
                        Other                          11.0
                        Product/Project Manager        13.0
                        Research Scientist             11.0
                        Software Engineer               8.0
                        Statistician                   10.0
                        Student                         NaN
Prefer not to say       Business Analyst               11.0
                        DBA/Database Engineer          13.0
                        Data Analyst                   11.0
                        Data Engineer                   9.0
                        Data Scientist                 14.0
                        Not employed                    NaN
                        Other                           7.0
                        Product/Project Manager        13.0
                        Research Scientist              8.0
                        Software Engineer              12.0
                        Statistician                    0.0
                        Student                         NaN
Prefer to self-describe Business Analyst                NaN
                        Data Engineer                  20.0
                        Data Scientist                 18.5
                        Not employed                    NaN
                        Other                          15.0
                        Research Scientist             10.0
                        Software Engineer               8.0
                        Statistician                    2.0
                        Student                         NaN

In [43]:
salary_job_gender["median salary range"] = ""
for idx in salary_job_gender.index:
  if salary_job_gender['salary map'][idx] == 4.5:
    salary_job_gender['salary map'][idx] = 5.0
  if salary_job_gender['salary map'][idx] == 18.5:
    salary_job_gender['salary map'][idx] = 19.0   
salary_job_gender["median salary range"] = salary_job_gender['salary map'].map(swap_salary_map)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [44]:
salary_job_gender

salary map median salary range
gender                  job title                                              
Female                  Business Analyst                7.0       10,000-14,999
                        DBA/Database Engineer          10.0       25,000-29,999
                        Data Analyst                    7.0       10,000-14,999
                        Data Engineer                   7.0       10,000-14,999
                        Data Scientist                 11.0       30,000-39,999
                        Not employed                    NaN                 NaN
                        Other                           7.0       10,000-14,999
                        Product/Project Manager        11.0       30,000-39,999
                        Research Scientist              8.0       15,000-19,999
                        Software Engineer               7.0       10,000-14,999
                        Statistician                    5.0         5,000-7,499
                        Student                         NaN                 NaN
Male                    Business Analyst                9.0       20,000-24,999
                        DBA/Database Engineer          11.0       30,000-39,999
                        Data Analyst                    8.0       15,000-19,999
                        Data Engineer                  11.0       30,000-39,999
                        Data Scientist                 12.0       40,000-49,999
                        Not employed                    NaN                 NaN
                        Other                          11.0       30,000-39,999
                        Product/Project Manager        13.0       50,000-59,999
                        Research Scientist             11.0       30,000-39,999
                        Software Engineer               8.0       15,000-19,999
                        Statistician                   10.0       25,000-29,999
                        Student                         NaN                 NaN
Prefer not to say       Business Analyst               11.0       30,000-39,999
                        DBA/Database Engineer          13.0       50,000-59,999
                        Data Analyst                   11.0       30,000-39,999
                        Data Engineer                   9.0       20,000-24,999
                        Data Scientist                 14.0       60,000-69,999
                        Not employed                    NaN                 NaN
                        Other                           7.0       10,000-14,999
                        Product/Project Manager        13.0       50,000-59,999
                        Research Scientist              8.0       15,000-19,999
                        Software Engineer              12.0       40,000-49,999
                        Statistician                    0.0               0-999
                        Student                         NaN                 NaN
Prefer to self-describe Business Analyst                NaN                 NaN
                        Data Engineer                  20.0     150,000-199,999
                        Data Scientist                 19.0     125,000-149,999
                        Not employed                    NaN                 NaN
                        Other                          15.0       70,000-79,999
                        Research Scientist             10.0       25,000-29,999
                        Software Engineer               8.0       15,000-19,999
                        Statistician                    2.0         2,000-2,999
                        Student                         NaN                 NaN

In [45]:
salary_job_gender = salary_job_gender[salary_job_gender['median salary range'].notna()]

In [46]:
salary_job_gender

salary map median salary range
gender                  job title                                              
Female                  Business Analyst                7.0       10,000-14,999
                        DBA/Database Engineer          10.0       25,000-29,999
                        Data Analyst                    7.0       10,000-14,999
                        Data Engineer                   7.0       10,000-14,999
                        Data Scientist                 11.0       30,000-39,999
                        Other                           7.0       10,000-14,999
                        Product/Project Manager        11.0       30,000-39,999
                        Research Scientist              8.0       15,000-19,999
                        Software Engineer               7.0       10,000-14,999
                        Statistician                    5.0         5,000-7,499
Male                    Business Analyst                9.0       20,000-24,999
                        DBA/Database Engineer          11.0       30,000-39,999
                        Data Analyst                    8.0       15,000-19,999
                        Data Engineer                  11.0       30,000-39,999
                        Data Scientist                 12.0       40,000-49,999
                        Other                          11.0       30,000-39,999
                        Product/Project Manager        13.0       50,000-59,999
                        Research Scientist             11.0       30,000-39,999
                        Software Engineer               8.0       15,000-19,999
                        Statistician                   10.0       25,000-29,999
Prefer not to say       Business Analyst               11.0       30,000-39,999
                        DBA/Database Engineer          13.0       50,000-59,999
                        Data Analyst                   11.0       30,000-39,999
                        Data Engineer                   9.0       20,000-24,999
                        Data Scientist                 14.0       60,000-69,999
                        Other                           7.0       10,000-14,999
                        Product/Project Manager        13.0       50,000-59,999
                        Research Scientist              8.0       15,000-19,999
                        Software Engineer              12.0       40,000-49,999
                        Statistician                    0.0               0-999
Prefer to self-describe Data Engineer                  20.0     150,000-199,999
                        Data Scientist                 19.0     125,000-149,999
                        Other                          15.0       70,000-79,999
                        Research Scientist             10.0       25,000-29,999
                        Software Engineer               8.0       15,000-19,999
                        Statistician                    2.0         2,000-2,999

In [47]:
np.array(salary_job_gender.loc["Female"]['salary map'])

array([ 7., 10.,  7.,  7., 11.,  7., 11.,  8.,  7.,  5.])

In [48]:
salary_job_gender.loc["Female"]['salary map'].index

Index(['Business Analyst', 'DBA/Database Engineer', 'Data Analyst',
       'Data Engineer', 'Data Scientist', 'Other', 'Product/Project Manager',
       'Research Scientist', 'Software Engineer', 'Statistician'],
      dtype='object', name='job title')

In [49]:
color_palette_list = ['#A5C8E4', '#C0ECCC', '#F9F0C1', '#F6A8A6']

layout = go.Layout(
    title='Salary Distribution by Gender for each Job Title',
    font=dict(color='#909090'),
    xaxis=dict(
        title='Job Title',
        titlefont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
        showticklabels=True,
        tickangle=-45,
        tickfont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
  	),
    yaxis=dict(
        title="Median Salary Range",
        titlefont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        ),
        tickmode = 'array',
        tickvals = np.array(list(swap_salary_map.keys())),
        ticktext = np.array(list(swap_salary_map.values())),
        tickangle=0,
        tickfont=dict(
            family='Arial, sans-serif',
            size=12,
            color='#909090'
        )
    )
)

fig = go.Figure(layout=layout)
fig.add_trace(go.Bar(
    x=np.array(salary_job_gender.loc["Female"]['salary map'].index),
    y=np.array(salary_job_gender.loc["Male"]['salary map']),
    name='Male',
    marker_color='#A5C8E4'
))
fig.add_trace(go.Bar(
    x=np.array(salary_job_gender.loc["Female"]['salary map'].index),
    y=np.array(salary_job_gender.loc["Female"]['salary map']),
    name='Female',
    marker_color='#C0ECCC'
))
fig.add_trace(go.Bar(
    x=np.array(salary_job_gender.loc["Female"]['salary map'].index),
    y=np.array(salary_job_gender.loc["Prefer not to say"]['salary map']),
    name='Prefer not to say',
    marker_color='#F9F0C1'
))

fig.add_trace(go.Bar(
    x=np.array(salary_job_gender.loc["Female"]['salary map'].index),
    y=np.array(salary_job_gender.loc["Prefer to self-describe"]['salary map']),
    name='Prefer to self-describe',
    marker_color='#F6A8A6'
))


# Here we modify the tickangle of the xaxis, resulting in rotated labels.
fig.update_layout(barmode='group', xaxis_tickangle=-45)
fig.show()

As you can see in all the jobs, females earn less than males. It is also interesting that respondents that have the label 'Prefer to self-describe' earn the most for some of the jobs, such as Business Analyst and Database Engineer. However, it could be because there are not many people in this category. For example, there is only 1 respondent in the 'Prefer to self-describe' category who is a Business Analyst. This will cause the results to be skewed and inaccurate.

##Visualise Gender Distribution in each Country

In [50]:
country_gender = df.groupby(['gender', 'country']).agg({'country':'count'})
country_gender

country
gender                  country                                                    
Female                  Algeria                                                  12
                        Argentina                                                13
                        Australia                                                44
                        Austria                                                   5
                        Bangladesh                                                7
...                                                                             ...
Prefer to self-describe Russia                                                    4
                        Spain                                                     1
                        Turkey                                                    1
                        United Kingdom of Great Britain and Northern Ir...        6
                        United States of America                                 13

[176 rows x 1 columns]

In [51]:
labels = country_gender.loc["Female"].index
percentages = np.array((country_gender.loc["Female"]["country"] / gender.Female.sum())*100)
colors_bar = ['#9fcc47', '#4ac9e2' , '#e045c1' , '#efcb5d' , '#f4a69f' , '#63c92c' , '#7da6e8' , '#ffc9b2' ,'#f7b7cb', '#510d91', '#b5520c', '#021982']
trace = go.Pie(labels=labels, 
               hoverinfo='label+percent', 
               values=percentages, 
               textposition='outside',
               marker=dict(colors=color_palette_list),
               rotation=90)

layout = go.Layout(
                    title="Country Distribution for Females",
                    font=dict(family='Arial', size=12, color='#909090'),
                    legend=dict(x=0.9, y=0.5)
                    )
data = [trace]
fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='Country Distribution for Females|')

In [52]:
labels = country_gender.loc["Male"].index
percentages = np.array((country_gender.loc["Male"]["country"] / gender.Male.sum())*100)
colors_bar = ['#9fcc47', '#4ac9e2' , '#e045c1' , '#efcb5d' , '#f4a69f' , '#63c92c' , '#7da6e8' , '#ffc9b2' ,'#f7b7cb', '#510d91', '#b5520c', '#021982']
trace = go.Pie(labels=labels, 
               hoverinfo='label+percent', 
               values=percentages, 
               textposition='outside',
               marker=dict(colors=color_palette_list),
               rotation=90)

layout = go.Layout(
                    title="Country Distribution for Males",
                    font=dict(family='Arial', size=12, color='#909090'),
                    legend=dict(x=0.9, y=0.5)
                    )
data = [trace]
fig = go.Figure(data=data, layout=layout)
py.iplot(fig, filename='Country Distribution for Males')

For both males and females, a large percentage of the respondents came from either India or the United States of America.

In conclusion, there is gender disparity in the tech industry, and women are still under-represented in this field. In addition, there is a salary disparity between males and females in the tech industry. We should encourage more girls to join the tech industry, by setting up more programs conducted by women already in the industry to inspire more girls to take up more STEM jobs.